llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp

```diff
@@ -1389,16 +1397,36 @@ bool DWARFDebugLine::LineTable::lookupAddressRangeImpl(
   Sequence.SectionIndex = Address.SectionIndex;
   Sequence.HighPC = Address.Address;
   SequenceIter LastSeq = Sequences.end();
-  SequenceIter SeqPos = llvm::upper_bound(
-      Sequences, Sequence, DWARFDebugLine::Sequence::orderByHighPC);
-  if (SeqPos == LastSeq || !SeqPos->containsPC(Address))
-    return false;
+  SequenceIter SeqPos;
+
+  if (StmtSequenceOffset) {
+    // If we have a statement sequence offset, find the specific sequence
+    // Binary search for sequence with matching StmtSeqOffset
+    Sequence.StmtSeqOffset = *StmtSequenceOffset;
+    SeqPos = std::lower_bound(Sequences.begin(), LastSeq, Sequence,
+                              [](const DWARFDebugLine::Sequence &LHS,
+                                 const DWARFDebugLine::Sequence &RHS) {
+                                return LHS.StmtSeqOffset < RHS.StmtSeqOffset;
+                              });
+
+    // If sequence not found or doesn't contain the address, return false
+    if (SeqPos == LastSeq || SeqPos->StmtSeqOffset != *StmtSequenceOffset ||
+        !SeqPos->containsPC(Address))
+      return false;
+
+    // Set LastSeq to just past this sequence since we only want this one
+    LastSeq = std::next(SeqPos);
```

Comment:
This is just a std::vector, so there's no need for the fancy std::next, this can be `LastSeq = SeqPos + 1;` (std::next is handy if you have a non-random-access container/iterator, etc, but otherwise I think it's easier to read with the more direct expression)